## Understanding Algorithms for Recommendation Systems

### * User-Product Relationships
#### Some people have a preference for specific products and this is measured via - User-Product Relationships.

### * Product-Product Relationships
#### Some products are similar in nature.
#### Examples:
##### - Books from the same genre;
##### - Dishes from the same cusinie;
##### - News articles about an event;

### * User-User Relationships
#### Some people are similar in nature.
##### They like the same books, common friends, similar age group, background;

### ---------------------------------------------------------------------------------------------

### Relationships among users and products
#### - These relationships can provide tremendenous insight:
#### What boooks a person will buy? If a person buys a phone what else will they buy? If A knows B, and B knows C, does A know C?
#### Relationships --------> Insight --------------> Monetization.
### Insight is Monetizable:
#### - Personalized promotional emails, homapages, notifications, feeds;

### ---------------------------------------------------------------------------------------------

### Identify Relationships

#### 1. Acquire a large amount of data. 
##### - User behavior data (Rating, Clicks, Purchases);
##### - User Demographics Data (Age, Education, Income, Location)
##### - Product Attribute Data (Genre-Books, Cast-Movies, Cuisine -Foods);

#### 2. Feed it to a system that can process it and extract the information about user/product relationships from it - Recommendation system;
##### -The recommendation system will use the algorithm to extract the relevant information from the data.

### ---------------------------------------------------------------------------------------------

### Types of Recommendation Algorithms

#### The purpose of recommendation algorithm is to extract information from data about what kinds of relationships might exist between users and products.
#### From the data user preferences can be identified and from the inference we could further recommend the similar products.

#### Recommendations can be done in various ways as listed
##### Option 1: Content based filtering - Find products with "similar" attributes;
##### Option 2: Collaborative filtering - Find products liked by "similar" users;
##### Option 3: Associate Rules Learning - Find "complementary" products; - Buy smartphone, use maybe likely to buy headphones;

### 1. Content Based Filtering:
#### Filtering based on similar attributes;
#### Example:
##### - User A -------------------- like ----- "Lord of the Rings"
##### - Algorithm would require a database that has ratings for Lord of the rings against different attributes.

In [1]:
# ----------------------------------
# "Lord of the Rings"    |  Rating
# ----------------------------------
#     Direction          |  10
# ----------------------------------
#      Cast              |  8
# ----------------------------------
#    Cinematography      |  10
# ----------------------------------
#         Story          |  9
# ----------------------------------

##### - The algorithm will look for other movies which have similar ratings against the same attributes. Example The Hobbit.

In [2]:
# --------------------------------
#  "The Hobbit"       |  Rating
# --------------------------------
#   Direction         |  9.5
# --------------------------------
#     Cast            |  8
# --------------------------------
#   Cinematography    |  9
# --------------------------------
#     Story           |  10
# --------------------------------

##### - Then the movie Hobbit can be recommended to User A.

### 2. Collabrative Filtering

#### User A likes the movie "Lord of the Rings" and "The Hobbit"
#### User B likes some of the same movies as User A and others. Example ("Lord of the Rings", "The Hobbit", "Casablanca");
#### User A and User B have common interests, hence movies that user B has liked can be recommended to user A.

### 3. Association Rules Learning
#### Find complementary products.
#### A user who buys smartphones is also likely to buy headphones.

### The Collaborative Filtering is the most common based recommendation algorithm;

### ---------------------------------------------------------------------------------------------

### Digging Dipper into Content Based Filtering
#### Content Based Filtering - Recommend products which have "similar" products.
### * What attributes should be considered?
##### - Identify attributes/factors atht influence user preferences. (Domain Expert)
##### - Movie (Direction, Cast, Cinematography, Story);
##### - Books (Author, Genre, Story);

### * How do we measure "similarity"?
##### - Rate the products against the chosen attributes.

In [3]:
#                          "Lord of the Rings" | "The Hobbit
# ----------------------------------------------------------
#     Direction          |  10                 | 9.5  
# ----------------------------------------------------------
#      Cast              |  8                  | 8   
# ----------------------------------------------------------
#    Cinematography      |  10                 | 9
# ----------------------------------------------------------
#         Story          |  9                  | 10
# ----------------------------------------------------------

#### The rating in points can be represented in N-dimensional space.
#### Given any pair of products they can be represented as points in N-dimensional space, once these points are plotted the distance between these points can be measured and this would give the proximity of the two products.
#### The proximity of the products in the N-dimensional space is the measure of similarity.
#### "Similarity" is measured using distance metrics. Example: - Euclidean distance, Hamming Distance, Correlation distance, etc;

### A Full-fledged Content Based Filtering System
#### User ------> Content-Based Filtering System ---------> Recommended Products.
#### * Rate every product against the relevant attributes.Each product is a number and can be represented in some N-dimensional space.

In [4]:
#       Product |    F1   | F2  | F3  | F4
# ----------------------------------------------------------
#       A       |  0     |  10  | 9.5 | 5
# ----------------------------------------------------------
#       B       |   5   |  8    | 8   | 4
# ----------------------------------------------------------
#       C       |   3   |  10   | 9  |  1
# ----------------------------------------------------------
#       D       |  9    |  9    | 10 |  2
# ----------------------------------------------------------

#### * Rate the user on the importance he/she gives to these factors. 
##### Example: Average of ratings of the products that the user already likes.
##### The user is assigned numbers and can be represented along the same axes as the products.
##### Then you can measure the distance between the user and every product in space and find the nearest  neighbors to the users, these are the recommended systems.

In [5]:
# ----------------------------------------------------------
#       A       |  0     |  10  | 9.5 | 5
# ----------------------------------------------------------

### ---------------------------------------------------------------------------------------------

### Digging Deeper into Collaborative Filtering

#### Content based filtering requires a product attribute database. 
#### Collaborative filtering uses easily captured user behavior data.
#### User's Affinity for products --------> Collaborative filtering ----------> Affinity for unseen products;
#### User's Affinity for products can be captured via - Purchases, Pageviews, Clicks, Ratings;
#### For any given user or product collaborative filtering algorithms require a single measure of affinity for that product by that user. (Purchases, Pageviews, Clicks, Ratings;)
#### Types of Rating 
##### 1. Explicit - direct -> survey, website;
##### 2. Implicit - indirect -> clicks, pageviews;
#### The ratings can be represented by a Rating Matrix - Each cell represents a rating.

In [6]:
#                      Rating Matrix
#         ----------------------------------
#         |      |  P1 | P2 | P3 | P4 | P5 |
#         ----------------------------------
#         |  U1  |  3  | 4  | -- | -- | -- |
#         ----------------------------------
#         |  U2  |  3  | 2  | -- | -- | 5  |
#         ----------------------------------
#         |  U3  |  -- | 2  | -- | 5  | 4  |
#         ----------------------------------
#         |  U4  |  -- | -- | 4  | -- | -- |
#         ----------------------------------
#         |  U5  |  1  | -- | -- | -- | -- |
#         ----------------------------------
#         |  U6  |  3  | 4  | -- | -- | 5  |

#### The existing blank values will be extrapolated using collabarative filtering, based on the filled cells the blank cell values can be predicted. 
#### Different ML techniques to fill the blanks:
##### 1. Nearest Neighbor model - Use the ratings of "most similar" users;
##### 2. Latent Fator Analysis - Solve for underlying factors that drive the ratings;

### ---------------------------------------------------------------------------------------------

### Contrasting Different Recommendation Algorithms:

### 1. Content Based Filtering:
#### - Find products with similar attributes.
#### - The products and users need to be mapped along attributes
#### - The users are products are represented in N-dimensional space;
#### - Centered around users.
#### - Problem: Intensive data garnering process; (A database with products rated against relevant attributes.) - cumbersome process/ not scalable;
#### - Music genome project;
#### 2 steps process:
##### a. Manual process with human intervension;
##### b. An algorithm to extract information;

### 2. Collaborative Filtering:
#### - Find products liked by "similar" users.
#### - Directly captured behavior data in form of ratings and represented as a matrix.
#### - It can be represented as users in an N dimensional space.
#### - Centered around users.
#### - Most Popular - Based on user behavior, agnostic to product attributes, no human intervension required.
#### - 1 step process:
##### a. Extract information directly from user rating;

### 3. Associate Rules Learning:
#### - Find "complementary" products.
#### - Data Representation - Data is not represented in vector form, but in the form of conditional probability - P(Headphones/ mobile phone);
#### - Purpose - find products associated with other products.

### ---------------------------------------------------------------------------------------------

### Recommending Products Based on Nearest Neighbor Model

#### Book Crossing - website for book reference rating.
#### Collaborative Filtering Techniques:
##### a. Nearest Neighbors Model - A new users rating is obtained from using the ratings of "most similar" users.
##### b. Latent Factor Analysis -  Take a matrix and solve for a set of underlying factors that drive the ratings responsible for the matrix ratings.

### Finding Top 10 Book Recommendations:
#### 1. Find the K nearest neighbors of a user.
#### 2. Average the ratings of the nearest neighbors for unrated books.
#### 3. Sort in descending order.
#### 4. Pick Top 10 books in the model.

### ---------------------------------------------------------------------------------------------

### Distance Metrics
#### 1. Euclidean Distance - measure distance between point (same as in 2D/3D);
#### 2. Correlation Distance - measure the correlation between points.
#### 3. Hamming Distance - measures how many numbers are in agreement between the two sets of numbers. 

### 1. Euclidean Distance:
#### Measuring the distance - pythagores distance = Sqrt((x2 - x1)^2 + (y2-y1)^2);
#### Preferred for continuous Data.

### 2. Colleration Distance:
#### Measure of the average rating;
#### U1 -> 1 4 3 4 2
#### U2 -> 3 4 2 4 4
#### Theses user v/s product ratings are depicted in a line graph, and a dotted average line is drwn as a metrics.
#### Outcomes:
#### For P2 - both users have given high rating; for P3 both users have given low rating; Correlation measures how in sync the variations or deviations from the mean for each of the rating;
#### U1 -> x1, x2,......xn; ---> mean xbar;
#### U2 -> y1, y2,......yn; ---> mean ybar;

#### Corr(x,y) = sum(xi - xbari)/ sqrt(sum(xi-xbar)^2) sqrt(sum(yi-ybar)^2)
#### Correlation is the measure of similarity; Lies in the range [-1, 1];
#### Correlation Distance = 1 - Correlation;
#### Preferred for continuous Data.

### 3. Hamming Distance
#### measures the percentage disagreement between two series of numbers.
#### U1 -> 1 4 3 4 2
#### U2 -> 3 4 2 4 4
#### The two users agree 40% (P2, P4) , disagree - 60% of the times;
#### Therefore, Disagreement = 0.6;
#### Preferred for discrete Data.

#### Implementing the Nearest Neighbors Model
#### 1. Set up the data - functions to access relevant information; 
##### * (Load 2 files - Ratings, Book Metadata);
##### * Function- lookup metadata for ISBN, function to find the favorite books of the user;

#### 2. Construct a rating matrix - The representation needed for collaborative filtering;
##### * From the table each row is a rating that will be converted into the Rating Matric - pandas.pivot_table;
##### * The arting matrix could be parse, they could be imputed, or restrict the size of the matrix for better computational performance;

#### 3. Find the K Nearest Neighbors 
##### create a function that can compute the distance between any pair of users - using the scipy lib;
##### the distnace function will use a active user and value K are will return the K nearest neighbors to this user.


#### 4. Find the top N recommendations - Use the average ratings of K nearest neighbors.
##### Average rhe rating of bearest neigbours for unrated books;
##### Sort in descending order;
##### Pick the top N;

### Nearest Neighbor Model

#### Step 1: Set up the data 
##### Functions to access relevant information

In [7]:
import pandas as pd
dataFile='C:/Users/AMI_1234/Documents/Computer_programs/ML/recommend_dataset/BX-Book-Ratings.csv'
data=pd.read_csv(dataFile,header=0,names =["user", "isbn", "rating"], sep=";", encoding='ISO-8859-1')

In [8]:
len(data)

1149780

In [9]:
data.head()

user        isbn  rating
0  276725  034545104X       0
1  276726  0155061224       5
2  276727  0446520802       0
3  276729  052165615X       3
4  276729  0521795028       6

##### Top rated books

In [10]:
bookFile = 'C:/Users/AMI_1234/Documents/Computer_programs/ML/recommend_dataset/BX-Books.csv'
books=pd.read_csv(bookFile,header=0,names =["isbn", "title", "author"], sep=";", encoding='ISO-8859-1', error_bad_lines=False, usecols=[0,1,2],index_col=0)

In [11]:
books.head()

title  \
isbn                                                            
0195153448                                Classical Mythology   
0002005018                                       Clara Callan   
0060973129                               Decision in Normandy   
0374157065  Flu: The Story of the Great Influenza Pandemic...   
0393045218                             The Mummies of Urumchi   

                          author  
isbn                              
0195153448    Mark P. O. Morford  
0002005018  Richard Bruce Wright  
0060973129          Carlo D'Este  
0374157065      Gina Bari Kolata  
0393045218       E. J. W. Barber

In [12]:
def bookMeta(isbn):
    title = books.at[isbn,"title"]
    author = books.at[isbn,"author"]
    return title, author
bookMeta("0002005018")

('Clara Callan', 'Richard Bruce Wright')

In [13]:
def favBooks(user,N):
    userRating = data[data["user"]==user]
    sortedRatings = pd.DataFrame.sort_values(userRating,['rating'],ascending=[0])[:N]
    sortedRatings["title"] = sortedRatings["isbn"].apply(bookMeta)
    return sortedRatings

In [14]:
data = data[data["isbn"].isin(books.index)]

In [15]:
favBooks(204622,5)

user        isbn  rating  \
844955  204622  0967560500      10   
844935  204622  0671027360      10   
844926  204622  0385504209      10   
844958  204622  097173660X       9   
844920  204622  0060935464       9   

                                                   title  
844955        (Natural Hormonal Enhancement, Rob Faigin)  
844935                  (Angels &amp; Demons, Dan Brown)  
844926                    (The Da Vinci Code, Dan Brown)  
844958  (Life After School Explained, Cap &amp; Compass)  
844920               (To Kill a Mockingbird, Harper Lee)

#### 2. Construct a rating matrix
##### The representation needed for collaborative filtering

In [16]:
data.shape

(1031175, 3)

#### Currently, there are 1031175 rows of data, with 3 columns - user, isbn, rating - this must be converted into a rating matrix with users listed along the y axis and isbn along the x-axis and the ratings must fit into the cell.

In [17]:
# Distinct isbn's and the number of occurences
usersPerISBN = data.isbn.value_counts()
usersPerISBN.head(10)

0971880107    2502
0316666343    1295
0385504209     883
0060928336     732
0312195516     723
044023722X     647
0142001740     615
067976402X     614
0671027360     586
0446672211     585
Name: isbn, dtype: int64

In [18]:
usersPerISBN.shape

(270170,)

In [19]:
# Distinct isbn's and the number of occurences
ISBNPeruser = data.user.value_counts()
ISBNPeruser.head(10)

11676     11144
198711     6456
153662     5814
98391      5779
35859      5646
212898     4290
278418     3996
76352      3329
110973     2971
235105     2943
Name: user, dtype: int64

In [20]:
ISBNPeruser.shape

(92107,)

In [21]:
# Due to large volume of data, consider ISBNs of books that are read by more 
# than 10 users.
data = data[data["isbn"].isin(usersPerISBN[usersPerISBN>20].index)]

In [22]:
# Due to large volume of data, consider users where ISBNs are read by more 
# than 10 users.
data = data[data["user"].isin(ISBNPeruser[ISBNPeruser>20].index)]

In [23]:
userItemRatingMatrix=pd.pivot_table(data, values='rating', index=['user'], columns=['isbn'])

In [24]:
userItemRatingMatrix.head()

isbn  000649840X  0006547834  0006550576  0006550789  0007110928  0007141076  \
user                                                                           
242          NaN         NaN         NaN         NaN         NaN         NaN   
243          NaN         NaN         NaN         NaN         NaN         NaN   
254          NaN         NaN         NaN         NaN         NaN         NaN   
383          NaN         NaN         NaN         NaN         NaN         NaN   
388          NaN         NaN         NaN         NaN         NaN         NaN   

isbn  0007154615  0020198817  0020198906  0020199600     ...      880781210X  \
user                                                     ...                   
242          NaN         NaN         NaN         NaN     ...             NaN   
243          NaN         NaN         NaN         NaN     ...             NaN   
254          NaN         NaN         NaN         NaN     ...             NaN   
383          NaN         NaN         NaN         NaN     ...             NaN   
388          NaN         NaN         NaN         NaN     ...             NaN   

isbn  8807813025  8817106100  8817106259  8817131628  8845205118  8845247414  \
user                                                                           
242          NaN         NaN         NaN         NaN         NaN         NaN   
243          NaN         NaN         NaN         NaN         NaN         NaN   
254          NaN         NaN         NaN         NaN         NaN         NaN   
383          NaN         NaN         NaN         NaN         NaN         NaN   
388          NaN         NaN         NaN         NaN         NaN         NaN   

isbn  884590184X  8885989403  950491036X  
user                                      
242          NaN         NaN         NaN  
243          NaN         NaN         NaN  
254          NaN         NaN         NaN  
383          NaN         NaN         NaN  
388          NaN         NaN         NaN  

[5 rows x 6863 columns]

#### 3. Finding Top N Recommendations

In [25]:
user1 = 204622
user2 = 255489

In [26]:
# pandas allows to pick a column from a data frame usig column name, 
# but not row, hence need to transpose it.
user1Ratings = userItemRatingMatrix.transpose()[user1]
user1Ratings.head()

isbn
000649840X   NaN
0006547834   NaN
0006550576   NaN
0006550789   NaN
0007110928   NaN
Name: 204622, dtype: float64

In [27]:
user2Ratings = userItemRatingMatrix.transpose()[user2]
user2Ratings.head()

isbn
000649840X   NaN
0006547834   NaN
0006550576   NaN
0006550789   NaN
0007110928   NaN
Name: 255489, dtype: float64

In [28]:
from scipy.spatial.distance import hamming
hamming(user1Ratings, user2Ratings)
# computes the percentage disagreement between the two users.

0.99985429112632962

In [29]:
import numpy as np
def distance(user1,user2):
    try:
        user1Ratings = userItemRatingMatrix.transpose()[user1]
        user2Ratings = userItemRatingMatrix.transpose()[user2]
        distance = hamming(user1Ratings, user2Ratings)
    except:
        distance = np.NaN
    return distance

In [30]:
distance(204622,255489)

0.99985429112632962

In [31]:
# active user, find the k nearest neighbours.
user = 204622
allUsers = pd.DataFrame(userItemRatingMatrix.index) # all user in dataframe
allUsers = allUsers[allUsers.user!=user] # Remove the active user

In [32]:
allUsers.head()

user
0   242
1   243
2   254
3   383
4   388

In [33]:
## apply the distance function
allUsers["distance"] = allUsers["user"].apply(lambda x: distance(user,x))

In [34]:
allUsers.head()

user  distance
0   242  0.999854
1   243  0.999854
2   254  1.000000
3   383  1.000000
4   388  1.000000

In [35]:
# Find the K-nearest neighbors
K = 10
KnearestUsers = allUsers.sort_values(["distance"],ascending=True)["user"][:K]

In [36]:
KnearestUsers

1490     68555
1821     82893
4431    198711
1930     87555
319      16795
136       7346
3131    140036
5693    251422
1050     48046
5176    232131
Name: user, dtype: int64

In [37]:
def nearestNeighbors(user,K=10):
    allUsers = pd.DataFrame(userItemRatingMatrix.index)
    allUsers = allUsers[allUsers.user!=user]
    allUsers["distance"] = allUsers["user"].apply(lambda x: distance(user,x))
    KnearestUsers = allUsers.sort_values(["distance"],ascending=True)["user"][:K]
    return KnearestUsers

In [38]:
KnearestUsers = nearestNeighbors(user)

In [39]:
KnearestUsers

1490     68555
1821     82893
4431    198711
1930     87555
319      16795
136       7346
3131    140036
5693    251422
1050     48046
5176    232131
Name: user, dtype: int64

#### 4. Finding the TOP N Recommendations
##### Finding the nearest neighbors for a user.

In [40]:
## gets the Rating of the nearest neighbors for all books for the given active user.
## Each column in NNRating represents a book.
NNRatings = userItemRatingMatrix[userItemRatingMatrix.index.isin(KnearestUsers)]
NNRatings

isbn    000649840X  0006547834  0006550576  0006550789  0007110928  \
user                                                                 
7346           NaN         NaN         NaN         NaN         NaN   
16795          NaN         NaN         NaN         NaN         NaN   
48046          NaN         NaN         NaN         NaN         NaN   
68555          NaN         NaN         NaN         NaN         NaN   
82893          NaN         NaN         NaN         NaN         NaN   
87555          NaN         NaN         NaN         NaN         NaN   
140036         NaN         NaN         NaN         NaN         NaN   
198711         NaN         NaN         NaN         NaN         NaN   
232131         NaN         NaN         NaN         NaN         NaN   
251422         NaN         NaN         NaN         NaN         NaN   

isbn    0007141076  0007154615  0020198817  0020198906  0020199600  \
user                                                                 
7346           NaN         NaN         NaN         8.0         NaN   
16795          NaN         NaN         NaN         NaN         NaN   
48046          NaN         NaN         NaN         NaN         NaN   
68555          NaN         3.0         NaN         NaN         NaN   
82893          NaN         NaN         NaN         NaN         NaN   
87555          NaN         0.0         NaN         NaN         0.0   
140036         NaN         NaN         NaN         8.0         NaN   
198711         NaN         NaN         NaN         NaN         NaN   
232131         NaN         NaN         NaN         NaN         NaN   
251422         NaN         NaN         0.0         NaN         NaN   

isbn       ...      880781210X  8807813025  8817106100  8817106259  \
user       ...                                                       
7346       ...             NaN         NaN         NaN         NaN   
16795      ...             NaN         NaN         NaN         NaN   
48046      ...             NaN         NaN         NaN         NaN   
68555      ...             NaN         NaN         NaN         NaN   
82893      ...             NaN         NaN         NaN         NaN   
87555      ...             NaN         NaN         NaN         NaN   
140036     ...             NaN         NaN         NaN         NaN   
198711     ...             NaN         NaN         NaN         NaN   
232131     ...             NaN         NaN         NaN         NaN   
251422     ...             NaN         NaN         NaN         NaN   

isbn    8817131628  8845205118  8845247414  884590184X  8885989403  950491036X  
user                                                                            
7346           NaN         NaN         NaN         NaN         NaN         NaN  
16795          NaN         NaN         NaN         NaN         NaN         NaN  
48046          NaN         NaN         NaN         NaN         NaN         NaN  
68555          NaN         NaN         NaN         NaN         NaN         NaN  
82893          NaN         NaN         NaN         NaN         NaN         NaN  
87555          NaN         NaN         NaN         NaN         NaN         NaN  
140036         NaN         NaN         NaN         NaN         NaN         NaN  
198711         NaN         NaN         NaN         NaN         NaN         NaN  
232131         NaN         NaN         NaN         NaN         NaN         NaN  
251422         NaN         NaN         NaN         NaN         NaN         NaN  

[10 rows x 6863 columns]

In [41]:
## Calaculate the Average Rating per book
avgRating = NNRatings.apply(np.nanmean).dropna()
avgRating.head()

C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeW

isbn
0007154615    1.5
0020198817    0.0
0020198906    8.0
0020199600    0.0
002026478X    0.0
dtype: float64

In [42]:
## Get the ratings from the active user - index - gives the isbn
booksAlreadyRead = userItemRatingMatrix.transpose()[user].dropna().index
booksAlreadyRead

Index(['006016848X', '0060935464', '0140042598', '0142004278', '0380732238',
       '0385504209', '0425109720', '0425152898', '0440136482', '0440241162',
       '0451191145', '0553096060', '0671027360', '0671027387', '0743225708',
       '076790592X', '0786868716', '0802132952', '0971880107', '1853260045',
       '1853260207', '185326041X', '1878424114'],
      dtype='object', name='isbn')

In [43]:
# Remove the average Ratings for books already read by user.
avgRating = avgRating[~avgRating.index.isin(booksAlreadyRead)]

In [44]:
# Sort ratings in descending order, these are the books not rated by the active 
# user but have high ratings from the nearest neighbors, pick Top N
N = 3
topNISBNs = avgRating.sort_values(ascending=False).index[:N]

In [45]:
pd.Series(topNISBNs).apply(bookMeta)

0    (The Curious Incident of the Dog in the Night-...
1          (The Autobiography of Malcolm X, Malcolm X)
2    (Demon Lord of Karanda (Malloreon (Paperback R...
Name: isbn, dtype: object

In [46]:
def topN(user,N=3):
    #find the nearest neighbors;
    KnearestUsers = nearestNeighbors(user) 
    # Get the ratings of the nearest neighbors for all books.
    NNRatings = userItemRatingMatrix[userItemRatingMatrix.index.isin(KnearestUsers)]
    # Find the average rating of the nearest neighbors for all books.
    avgRating = NNRatings.apply(np.nanmean).dropna() 
    # get all the books in the dataset;
    booksAlreadyRead = userItemRatingMatrix.transpose()[user].dropna().index
    # Remove the average ratings for books already rated by the active user.
    avgRating = avgRating[~avgRating.index.isin(booksAlreadyRead)]
    # Sort the average ratings in descending order, pick the top N books based on 
    # the nearest neighbor's rating.
    topNISBNs = avgRating.sort_values(ascending=False).index[:N]
    # apply the book metadata function to get the Top N ISBNs.
    return pd.Series(topNISBNs).apply(bookMeta)

In [47]:
# Print the list of favorite books for a given user
favBooks(204813, 10)

user        isbn  rating  \
845417  204813  0399149848      10   
845407  204813  0385504209      10   
845359  204813  0142001805      10   
845416  204813  0399149392      10   
845382  204813  0373218036      10   
845431  204813  0446527793      10   
845442  204813  051513628X       9   
845451  204813  0671027360       9   
845433  204813  0446532452       9   
845434  204813  0446606243       9   

                                                    title  
845417                         (Birthright, Nora Roberts)  
845407                     (The Da Vinci Code, Dan Brown)  
845359          (The Eyre Affair: A Novel, Jasper Fforde)  
845416  (Chesapeake Blue (Quinn Brothers (Hardcover)),...  
845382             (Truly, Madly Manhattan, Nora Roberts)  
845431                    (The Guardian, Nicholas Sparks)  
845442  (Key of Light (Key Trilogy (Paperback)), Nora ...  
845451                   (Angels &amp; Demons, Dan Brown)  
845433                     (The Wedding, Nicholas Sparks)  
845434                  (The Tenth Justice, Brad Meltzer)

In [48]:
# top recommendations for that user
topN(204813, 10)

C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeWarning)
C:\Users\AMI_1234\Anaconda3\lib\site-packages\numpy\lib\nanfunctions.py:703: RuntimeWarning: Mean of empty slice
  warnings.warn("Mean of empty slice", RuntimeW

0                        (The Outsiders, S. E. Hinton)
1                   (Lethal Seduction, Jackie Collins)
2    (Waiting For Nick (Silhouette Special Edition)...
3                    (Range of Motion, Elizabeth Berg)
4    (Someone to Watch Over Me : A Novel, Judith Mc...
5    (HITCHHIK GD GALAXY (Hitchhiker's Trilogy (Pap...
6               (Sunset in St. Tropez, Danielle Steel)
7          (La Cucina: A Novel of Rapture, Lily Prior)
8                      (The Visitation, Frank Peretti)
9    (Little Women (Signet Classic), Louisa May Alc...
Name: isbn, dtype: object

## ------------------------------------------------------------------------- 

### Recommending Products Based on the Latent Factors Model

#### Latent Factor Analysis - inspired from the PCA (Principal Component Analysis) - In this technique a dataset is taken and then figure out the primary factors/principal components that explain the patterns.  
#### In Latent factor analysis - Driving conditions could be - Why certain products are rated high and the others low? Why certain users like certain products? what are the underlying factors that influence the user ratings?
#### Once the underlying factors are known, any product rating by a user can be predicted;

### Latent Factor Analysis Working:
#### 1. The rating matrix is decomposed into 2 separate matrices, one specific to user/factor(users - y-axis) and the other product/factor(products-x-axis); 
#### 2. Solve using the known ratings.
#### 3. Use the solution to find the unknown ratings.

## ------------------------------------------------------------------------- 

### Nearest Neighbors Model vs Latent Factor Analysis

#### 1. Data Representation:
##### Nearest Neighbor Model, the users are represented in N dimensional space, and the distance is found between users (observal attributes);
##### Laten Factor Analysis - users are also represented in an N dimensional space, but this representation must be transformed into new representation, where the dimensions are the driving factors for the rating (hidden driving factor).
#### 2. Model Update - online v/s offline updates
#### 3. Memory Usage - In-memory v/s pre-computed

##### ---------------------------------------------------------------------

#### Decomposing the Rating Matrix 
#### 1. Rating Matrix -> R ----> User matrix - P and  Product Matrix - Q
#### r32 = p3 * q2; rui = pu . qi; 
##### (pu - row corresponding to user in the matrix P and qi will be a column corresponding to an item in the matrix Q).
##### #Equation = #known ratings

#### Because of the unfilled ratings whatever is obtained for pu and qi will not be the final ratings , there will be a slight error, ei = min (sum(rui - pu-qi)^2);
#### Last adjustment - Add a term to penalize the model for the number of factors.
#### Regularization term , regularization factor (lambda)
### ei = min (sum(rui - pu-qi)^2) + lambda(||pu||^2+||qi||^2);

### Step 2: Construct a rating matrix

In [49]:
from scipy.sparse import coo_matrix
# Row source and column source must contain only integer values
data['user'] = data['user'].astype('category')
data['isbn'] = data['isbn'].astype('category')

R = coo_matrix((data['rating'].astype(float),
                       (data['user'].cat.codes.copy(),
                        data['isbn'].cat.codes.copy())))

In [50]:
data.head()

user        isbn  rating
173  276847  0446364193       0
175  276847  3379015180       0
210  276847  3551551677      10
413  276925  002542730X      10
414  276925  0060520507       0

In [51]:
data.shape

(275654, 3)

In [52]:
R.shape

(6290, 6863)

In [53]:
len(R.data)

275654

### Step 3: Initialize Factor Matrices

In [54]:
# M - users; N - isbns; K = factors
M,N = R.shape
K = 3

In [55]:
import numpy as np
P = np.random.rand(M,K) # users and factors
Q = np.random.rand(K,N) # factors and products

### Step 4: Compute the error

In [56]:
from numpy.linalg import norm
# lambda regularization factor - penalizing term
def error(R,P,Q,lamda=0.02):
    ratings = R.data
    rows = R.row
    cols = R.col
    e = 0
    for ui in range(len(ratings)):
        rui=ratings[ui]
        u = rows[ui]
        i = cols[ui]
        if rui>0:
            e= e + pow(rui-np.dot(P[u,:],Q[:,i]),2)+\
            lamda*(pow(norm(P[u,:]),2)+pow(norm(Q[:,i]),2))
    return e

In [57]:
error(R,P,Q)

4709825.9899968365

In [58]:
# Root Mean Square Error
rme = np.sqrt(error(R,P,Q)/len(R.data))

In [59]:
rme

4.1335222672698837

### Step 5: Use Stochastic Gradient Descent 
##### update the factor matrices and minimize the errors

#### K = #factors; lambda - regularization factor; steps = # max no of steps allowed; gamma = step size; 

In [60]:
def SGD(R, K, lamda=0.02, steps=10, gamma=0.001):
    
    M,N = R.shape
    P = np.random.rand(M,K)
    Q = np.random.rand(K,N)
    
    rmse = np.sqrt(error(R,P,Q,lamda)/len(R.data))
    print("Initial RMSE: "+str(rmse))
    
    for step in range(steps):
        for ui in range(len(R.data)):
            rui=R.data[ui]
            u = R.row[ui]
            i = R.col[ui]
            if rui>0:
                eui=rui-np.dot(P[u,:],Q[:,i])
                P[u,:]=P[u,:]+gamma*2*(eui*Q[:,i]-lamda*P[u,:])
                Q[:,i]=Q[:,i]+gamma*2*(eui*P[u,:]-lamda*Q[:,i])
        rmse = np.sqrt(error(R,P,Q,lamda)/len(R.data))
        if rmse<0.5:
            break
    print("Final RMSE: "+str(rmse))
    return P,Q

In [61]:
(P,Q)=SGD(R, K=2,gamma=0.0007,lamda=0.01, steps=2)

Initial RMSE: 4.26102213295
Final RMSE: 3.55637410748


In [62]:
(P,Q)=SGD(R, K=2,gamma=0.0007,lamda=0.01, steps=100)

Initial RMSE: 4.26111536729
Final RMSE: 0.779146026797


## ------------------------------------------------------------------------- 

### Understanding Association Rules
#### There are three types of Recommendation Algorithms
#### 1. Content Based Filtering - Find products with "similar" attributes;
#### 2. Collaborative Filtering - Find products liked by "similar" users;
#### 3. Associate Rules Learning - Find "complementary" products.

#### Association Rules Learning tries to identify the group of items that occur together often. What items are brought together in a trransaction? What items are bought by a user in a short period of item?
#### Associate Rules Learning is also termed as Market basket analysis.
#### The data representation here is interms of conditional probabilities. Analyis is done to figure out how the purchase of product B is impacted based on the purchase of product A. Does a person buying smartphone increase the likelihood of  buying headphones? (Directed)


### ---------------------------------

### Measuring the Strength of a Rule
#### 1. Support - measures how often a product or a group of products are sold independently.
#### 2. Confidence - measures the likelihood of products being sold together. (conditional probility that product B will be bought given that product A has been bought). It is the support of the products (A,B)/ support(A)
#### 3. Lift - Delta between Support and Confidence.Lift = Confidence/support;
#### It must have high support, cofidence and lift > 1 to be valid;

### ---------------------------------

### Mining for Rules Using the Apriori Algorithm
#### 1. Brute force technique - Find all the possible item sets and then compute the support and confidence measures for the rules that are generated and see which ones are valid - computationally very extensive.
#### 2. Apriori Algorithm - prune the number of items in each stage, use metrics to check how important as item set is (Support, Confidence);

#### Steps in Apriori Algorithm
##### 1. Find 1 item sets - Keep only those with minimum support
##### 2. Find 2 item sets - Use only the items left from previous step, keep only item sets with minimum support.
##### 3. Generate 2 item rules - Keep only those with minimum confidence.
##### 4. Find 3 item sets - Use only the items left from previous step, keep only item sets with minimum support.
##### 5. Generate 3 item rules - Keep only those with minimum confidence.
##### till no more items are left.

### ----------------------------------
### Association Rules for Bakery Items
#### 1. Set up the data - Receipts and item meta data.
#### 2. Compute the support - Set up a function to compute support for any items, then Confidence computation.
#### 3. Implement the Apriori Algorithm.

#### Dataset link - https://wiki.csc.calpoly.edu/datasets/wiki/ExtendedBakery
#### Dataset link - https://wiki.csc.calpoly.edu/datasets/wiki/ExtendedBakery75K
#### Download - 75000-out1.csv; EB-build-goods.sql;

#### 1. Set up the data - Receipts and item meta data

In [63]:
itempath = "C:/Users/AMI_1234/Documents/Computer_programs/ML/recommend_dataset/bakery_Items/EB-build-goods.sql"

In [64]:
receiptspath = "C:/Users/AMI_1234/Documents/Computer_programs/ML/recommend_dataset/bakery_Items/75000-out1.csv"

In [65]:
# read the data from the receipt file
with open(receiptspath,"r") as receiptsfile:
    receiptsdata = receiptsfile.read().split('\n')

In [66]:
# receipt mo, followed by items in the receipt
receiptsdata

['1, 11, 21',
 '2, 7, 11, 37, 45',
 '3, 3, 33, 42',
 '4, 5, 12, 17, 47',
 '5, 6, 18, 42',
 '6, 2, 4, 34',
 '7, 15, 16, 23, 40',
 '8, 2, 3, 29, 34',
 '9, 18, 23, 26, 35, 36',
 '10, 44, 45',
 '11, 17, 38, 48, 49',
 '12, 2, 3, 11, 21, 37, 41, 49',
 '13, 3, 17, 43, 48',
 '14, 17, 35, 43, 45',
 '15, 15, 37, 43',
 '16, 0, 2, 20, 46, 48',
 '17, 17, 47',
 '18, 14',
 '19, 16, 39',
 '20, 13, 42',
 '21, 7, 11, 37, 45',
 '22, 7, 15, 49',
 '23, 23, 24, 40, 41, 43',
 '24, 9, 15, 28, 47',
 '25, 32, 33, 37',
 '26, 5, 8, 16, 19, 20, 25, 39, 45',
 '27, 13, 22, 24, 32, 33',
 '28, 14, 44',
 '29, 6, 13, 20, 39, 40, 44, 49',
 '30, 13, 46',
 '31, 8, 27, 28',
 '32, 1, 19',
 '33, 6, 36',
 '34, 7, 15, 49',
 '35, 12, 31, 36, 48',
 '36, 17, 29, 47',
 '37, 5, 10, 21, 34, 37, 48',
 '38, 1, 19',
 '39, 4, 9',
 '40, 7, 15, 40, 49',
 '41, 18, 19, 21, 35',
 '42, 3, 17, 19',
 '43, 13, 14, 27',
 '44, 14, 19',
 '45, 0, 2, 6, 46',
 '46, 14, 30, 39, 44',
 '47, 0, 26, 46',
 '48, 0, 2, 46',
 '49, 1, 3, 12, 24, 31, 36, 39, 48',

In [67]:
# parse the receipt list and drop the items into the basket, remove receipt no.
baskets = [line.split(", ")[1:] for line in receiptsdata[0:-1]]

In [68]:
baskets[0:5]

[['11', '21'],
 ['7', '11', '37', '45'],
 ['3', '33', '42'],
 ['5', '12', '17', '47'],
 ['6', '18', '42']]

In [69]:
# read the data from the items file
with open(itempath,"r") as itemfile:
    lines = itemfile.read().split('\n')

In [70]:
# sql file
lines

["insert into goods values (0,'Chocolate','Cake',8.95,'Food');",
 "insert into goods values (1,'Lemon','Cake',8.95,'Food');",
 "insert into goods values (2,'Casino','Cake',15.95,'Food');",
 "insert into goods values (3,'Opera','Cake',15.95,'Food');",
 "insert into goods values (4,'Strawberry','Cake',11.95,'Food');",
 "insert into goods values (5,'Truffle','Cake',15.95,'Food');",
 "insert into goods values (6,'Chocolate','Eclair',3.25,'Food');",
 "insert into goods values (7,'Coffee','Eclair',3.5,'Food');",
 "insert into goods values (8,'Vanilla','Eclair',3.25,'Food');",
 "insert into goods values (9,'Napoleon','Cake',13.49,'Food');",
 "insert into goods values (10,'Almond','Tart',3.75,'Food');",
 "insert into goods values (11,'Apple','Pie',5.25,'Food');",
 "insert into goods values (12,'Apple','Tart',3.25,'Food');",
 "insert into goods values (13,'Apricot','Tart',3.25,'Food');",
 "insert into goods values (14,'Berry','Tart',3.25,'Food');",
 "insert into goods values (15,'Blackberry','T

In [71]:
items = [line.split("(")[1][0:-2].split(",") for line in lines[0:-1]]
items[0:5]

[['0', "'Chocolate'", "'Cake'", '8.95', "'Food'"],
 ['1', "'Lemon'", "'Cake'", '8.95', "'Food'"],
 ['2', "'Casino'", "'Cake'", '15.95', "'Food'"],
 ['3', "'Opera'", "'Cake'", '15.95', "'Food'"],
 ['4', "'Strawberry'", "'Cake'", '11.95', "'Food'"]]

In [72]:
itemMap = {line[0]:line[1]+" "+line[2] for line in items}

#### 2. Compute the support - Set up a function to compute support for any items, then Confidence computation

In [73]:
numItems = len(items)
numBaskets = len(baskets)

In [74]:
numItems

50

In [75]:
numBaskets

75000

In [76]:
# support(itemset) = no baskets will all items/ no total baskets
def support(itemset):
    # start will all baskets
    basketSubset = baskets 
    for item in itemset:
        # keep only baskets that contain that item
        basketSubset = [basket for basket in basketSubset if item in basket]
    return float(len(basketSubset))/float(numBaskets)

In [77]:
support(['2', '24'])
# Of all transactions only 0.28% of transactions contain 2 and 24.

0.00288

#### 3.Implement the Apriori Algorithm

#### a. Find 1 item sets - Keep only those with minimum support

In [78]:
# empty list to store the leftover items after iteration
supportItems1 = []
# custom minimum support
minsupport=0.01
# Iterate through all 1 item sets
for item in range(numItems):
    itemset=[str(item)]
    # keep only those item sets which have minimum support
    if support(itemset)>=minsupport:
        supportItems1.append(str(item))

In [79]:
supportItems1

['0',
 '1',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31',
 '32',
 '33',
 '34',
 '35',
 '36',
 '37',
 '38',
 '39',
 '40',
 '41',
 '42',
 '43',
 '44',
 '45',
 '46',
 '47',
 '48',
 '49']

In [80]:
# parameters - i - size of item sets in this iteration, 
# supportItems - items leftover from previous iteration;
# assocRules - Rules generated uptil now
# minsupport, minconfidence, and 
# an empty list to keep the left over items from current iteration.
#def aprioriIteration(i, supportItems, assocRules, newSupportItems, minsupport, minconfidence):
    # generate all item sets of size i,and keep those above minsupport
#     for itemset in itertools.combinations(supportItems, i):
#        itemset = list(itemset)
#        if support(itemset)>minsupport:
#            # generate rules from each item set
#            for j in range(i):
#                rule_to = itemset[j]
#                rule_from = [x for x in itemset if x!=itemset[j]]
#                confidence=support(itemset)/support(rule_from)
#                if confidence>minconfidence:
#                    assocRules.append((rule_from,rule_to))
#                    for x in itemset:
#                        if x not in newSupportItems:
#                            newSupportItems.append(x)
#    return assocRules, newSupportItems



In [81]:
import itertools
def aprioriIteration(i, supportItems, assocRules, newSupportItems, minsupport, minconfidence):
    
    for itemset in itertools.combinations(supportItems, i):
        itemset = list(itemset)
        if support(itemset)>minsupport:
            for j in range(i):
                rule_to = itemset[j]
                rule_from = [x for x in itemset if x!=itemset[j]]
                confidence=support(itemset)/support(rule_from)
                if confidence>minconfidence:
                    assocRules.append((rule_from,rule_to))
                    for x in itemset:
                        if x not in newSupportItems:
                            newSupportItems.append(x)
    return assocRules, newSupportItems



In [88]:
minsupport=0.01
minconfidence=0.5
assocRules=[]
newSupportItems=[]

assocRules, supportItems2 = aprioriIteration(2,supportItems1,assocRules,newSupportItems,minsupport,minconfidence)

In [89]:
assocRules

[(['46'], '0'),
 (['0'], '46'),
 (['3'], '18'),
 (['3'], '35'),
 (['9'], '4'),
 (['5'], '22'),
 (['44'], '14'),
 (['32'], '16'),
 (['16'], '32'),
 (['35'], '18'),
 (['18'], '35'),
 (['28'], '27'),
 (['27'], '28'),
 (['33'], '42')]

In [90]:
assocRules, newSupportItems3 = aprioriIteration(3,supportItems2,assocRules,newSupportItems,minsupport,minconfidence)

In [91]:
assocRules

[(['46'], '0'),
 (['0'], '46'),
 (['3'], '18'),
 (['3'], '35'),
 (['9'], '4'),
 (['5'], '22'),
 (['44'], '14'),
 (['32'], '16'),
 (['16'], '32'),
 (['35'], '18'),
 (['18'], '35'),
 (['28'], '27'),
 (['27'], '28'),
 (['33'], '42'),
 (['18', '35'], '3'),
 (['3', '35'], '18'),
 (['3', '18'], '35')]

In [92]:
def ruleMeta(rule):
    rule_from = [itemMap[x] for x in rule[0]]
    return rule_from,itemMap[rule[1]]

In [93]:
[ruleMeta(rule) for rule in assocRules]

[(["'Chocolate' 'Coffee'"], "'Chocolate' 'Cake'"),
 (["'Chocolate' 'Cake'"], "'Chocolate' 'Coffee'"),
 (["'Opera' 'Cake'"], "'Cherry' 'Tart'"),
 (["'Opera' 'Cake'"], "'Apricot' 'Danish'"),
 (["'Napoleon' 'Cake'"], "'Strawberry' 'Cake'"),
 (["'Truffle' 'Cake'"], "'Gongolais' 'Cookie'"),
 (["'Bottled' 'Water'"], "'Berry' 'Tart'"),
 (["'Apricot' 'Croissant'"], "'Blueberry' 'Tart'"),
 (["'Blueberry' 'Tart'"], "'Apricot' 'Croissant'"),
 (["'Apricot' 'Danish'"], "'Cherry' 'Tart'"),
 (["'Cherry' 'Tart'"], "'Apricot' 'Danish'"),
 (["'Tuile' 'Cookie'"], "'Marzipan' 'Cookie'"),
 (["'Marzipan' 'Cookie'"], "'Tuile' 'Cookie'"),
 (["'Cheese' 'Croissant'"], "'Orange' 'Juice'"),
 (["'Cherry' 'Tart'", "'Apricot' 'Danish'"], "'Opera' 'Cake'"),
 (["'Opera' 'Cake'", "'Apricot' 'Danish'"], "'Cherry' 'Tart'"),
 (["'Opera' 'Cake'", "'Cherry' 'Tart'"], "'Apricot' 'Danish'")]